pkg/fqdn/dnsproxy/shared_client.go

```diff
@@ -50,29 +50,14 @@ func (s *SharedClients) ExchangeContext(ctx context.Context, key string, conf *d
 func (s *SharedClients) ShutdownClient(key string) {
 	// lock for s.clients access
 	s.lock.Lock()
-	// locate client to re-use if possible.
 	client := s.clients[key]
 	if client == nil {
+		s.lock.Unlock()
 		return
 	}
+	delete(s.clients, key)
 	s.lock.Unlock()
-
-	client.Lock()
-	defer client.Unlock()
-	client.refcount--
-	if client.refcount == 0 {
-		// connection close must be completed while holding the client's lock to
-		// avoid a race where a new client dials using the same 5-tuple and gets a
-		// bind error.
-		// The client remains findable so that new users with the same key may wait
-		// for this closing to be done with.
-		client.close()
-		// Make client unreachable
-		// Must take s.lock for this.
-		s.lock.Lock()
-		delete(s.clients, key)
-		s.lock.Unlock()
-	}
+	client.close()
```

Comment:
We ought to lock the client when closing, no?

Reply:
So I thought about this for a bit, and I don't see why we'd need to. But to be honest, I think it's playing with fire so I'll just lock it and be done with it 😅